lib/Basic/SourceLoc.cpp

```diff
@@ -83,6 +83,25 @@ unsigned SourceManager::addMemBufferCopy(StringRef InputData,
   return addNewSourceBuffer(std::move(Buffer));
 }
 
+void SourceManager::createVirtualFile(SourceLoc Loc, StringRef Name,
+                                      int LineOffset, unsigned Length) {
+  CharSourceRange Range = CharSourceRange(Loc, Length);
+
+  // Skip if this range has already been added
+  VirtualFile &File = VirtualFiles[Range.getEnd().Value.getPointer()];
+  if (File.Range.isValid())
```

Comment:
Should we assert that the range and name are the same?

Reply:
Yep 👍